File: 54. Spiral Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        up = 0
        left = 0
        right = len(matrix[0]) -1
        bot = len(matrix) - 1
        result = []

        while len(result) < len(matrix[0]) * len(matrix):
            # traverse from left to right, row fixed, column changing.
            for i in range(left, right+1):
                result.append(matrix[up][i])
            
            # traverse from top to down. row changing, column fixed.
            for i in range(up+1, bot+1):
                result.append(matrix[i][right])
                
            #traverse from right to left. Make sure we only traverse
            # the row we haven't been. That means we only go from right to left
            # when up and bot haven't become the same row.
        
            if up != bot:
                for i in range(right-1, left-1, -1):
                    result.append(matrix[bot][i])
            
            #travere from bot to up. Make sure we only traverse the column
            # we haven't been. That means we only go from bot to top when
            # left and right haven't become the same column.

            if left != right:
                for i in range(bot-1, up, -1):
                    result.append(matrix[i][left])
                    
            #update the pointers,, make sure they squeeze the rectangle
            up += 1
            left  += 1
            right -= 1
            bot -= 1
        return result

File: 54. Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int up = 0;
        int left = 0;
        int right = matrix[0].length-1;
        int down = matrix.length-1;
        ArrayList<Integer> result = new ArrayList<Integer>();
        
        while (result.size() != matrix[0].length * matrix.length) {
            
            // go to right
            for (int l = left; l< right+1; l++){
                result.add(matrix[up][l]);
            }
            
            // go to down
            
            for (int u = up+1; u < down+1; u++){
                result.add(matrix[u][right]);
            }
            
            if (down != up){
                for (int r = right-1; r>left-1; r--){
                    result.add(matrix[down][r]);
                }
            }
            
            if (left != right){
                
                for (int d = down-1; d > up; d--){
                    result.add(matrix[d][left]);
                }
            }
            up++;
            down--;
            left++;
            right--;
            
            
        }
        return result;
        
    }
}
